pandevice/predefined.py

```diff
@@ -0,0 +1,269 @@
+#!/usr/bin/env python
+
+# Copyright (c) 2015, Palo Alto Networks
+#
+# Permission to use, copy, modify, and/or distribute this software for any
+# purpose with or without fee is hereby granted, provided that the above
+# copyright notice and this permission notice appear in all copies.
+#
+# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+# Author: John Anderson <lampwins@gmail.com>
+
+"""Retrieving and parsing predefined objects from the firewall"""
+
+from pandevice import getlogger
+import pandevice.errors as err
+from pan.xapi import PanXapiError
+from pandevice.updater import PanOSVersion
+import objects
+
+logger = getlogger(__name__)
+
+
+class Predefined(object):
+    """Predefined Objects Subsystem of Firewall
+
+    A member of a base.PanDevice object that has special methods for
+    interacting with the predefned objects of the firewall
+
+    This class is typically not instantiated by anything but the
+    base.PanDevice class itself. There is an instance of this Predefined class
+    inside every instantiated base.PanDevice class.
+
+    Args:
+        device (base.PanDevice): The firewall or Panorama this Predefined subsystem leverages
+
+    """
+
+    # xpath
+    PREDEFINED_ROOT_XPATH = "/config/predefined"
+    ENTRY = "/entry[@name='%s']"
+    SERVICE = "/service"
+    # apps and containers share a namespace so we need to search both
+    APPLICATION_CONTAINS_XPATH = '//*[contains(local-name(), "application")]'
+    ALL_APPLICATION_XPATH = PREDEFINED_ROOT_XPATH + APPLICATION_CONTAINS_XPATH
+    SINGLE_APPLICATION_XPATH = ALL_APPLICATION_XPATH + ENTRY
+    ALL_SERVICE_XPATH = PREDEFINED_ROOT_XPATH + SERVICE
+    SINGLE_SERVIE_XPATH = ALL_SERVICE_XPATH + ENTRY
+
+    def __init__(self, device=None, *args, **kwargs):
+        # Create a class logger
+        self._logger = getlogger(__name__ + "." + self.__class__.__name__)
+
+        self.parent = device
+
+        self.service_objects = {}
+        self.application_objects = {}
+        self.application_container_objects = {}
+
+    def _get_xml(self, xpath):
+        """use the parent to get the xml given the xpath"""
+
+        err_msg = "Predefined object(s) does not exist with xpath: {0}".format(xpath)
+
+        try:
+            root = self.parent.xapi.get(xpath, retry_on_peer=self.parent.HA_SYNC)
+        except (PanXapiError, err.PanNoSuchNode) as e:
+            raise err.PanObjectMissing(err_msg, pan_device=self.parent)
```

Comment:
I don't believe this is necessary to catch exceptions here.  It's catching these exceptions to raise another exception, and I don't think these exceptions will ever happen anyway.  For example, asking for an app that doesn't exist just return None, it doesn't raise an exception.

Is there a reason this is a try/except block?  Maybe remove try/except block and just let any exceptions fall through instead?